### PR TITLE
Ignore SIGPIPE events

### LINKED ISSUE #19728: journald restart crashes Docker
It appears that a journald restart on Redhat 7.2 crashes Docker. I was getting my logging settings in place when I noticed that after restarting journald my Docker daemon would crash. I backed things out to where even without any containers if I restarted journald Docker crashes and then restarts. 

It is very easy to reproduce. Just restart journald.

```
$ systemctl restart  systemd-journald.service
```

Then check Docker processes. After just a few times you get this.

```
$ docker ps
An error occurred trying to connect: Get http:///var/run/docker.sock/v1.21/containers/json: EOF
$ docker ps
Error response from daemon: 404 page not found
$ docker ps
CONTAINER ID        IMAGE ...
```

You can see it recovers. However, if I have containers running it will stop all of them and even require me to manually remove contatiners that show up as Dead. I also have to remove the containers sometimes as they appear to be in a state where they will not start without errors.

`docker version`:

Client:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   a34a1d5
 Built:        Fri Nov 20 13:25:01 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   a34a1d5
 Built:        Fri Nov 20 13:25:01 UTC 2015
 OS/Arch:      linux/amd64

`docker info`:

Containers: 13
Images: 62
Server Version: 1.9.1
Storage Driver: devicemapper
 Pool Name: docker-253:0-67131623-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 107.4 GB
 Backing Filesystem: 
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 4.038 GB
 Data Space Total: 107.4 GB
 Data Space Available: 31.62 GB
 Metadata Space Used: 5.886 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.142 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.107-RHEL7 (2015-10-14)
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 3.10.0-327.el7.x86_64
Operating System: Red Hat Enterprise Linux Server 7.2 (Maipo)
CPUs: 2
Total Memory: 7.64 GiB
Name: vmtstcon01.roomandboard.com
ID: 7J2R:VPDV:RZRX:NM3X:6JG3:5HGD:ZUIW:FGMW:RCJO:NULN:FS7Z:UD42
WARNING: bridge-nf-call-ip6tables is disabled

`uname -a`:

Linux vmtstcon01.roomandboard.com 3.10.0-327.el7.x86_64 #1 SMP Thu Oct 29 17:29:29 EDT 2015 x86_64 x86_64 x86_64 GNU/Linux

Environment details (AWS, VirtualBox, physical, etc.):

vmware

### PR DESCRIPTION
**Pull request by jwhonce**
Using golang 1.6, is it now possible to ignore SIGPIPE events on stdout/stderr. 
Previous versions of the golang library cached 10 events and then killed the 
process receiving the events.

systemd-journald sends SIGPIPE events when jounald is restarted and
the target of the unit file writes to stdout/stderr. Docker logs to stdout/stderr.

This patch silently ignores all SIGPIPE events.

This resolves #19728

To reproduce issue:
- systemctl start docker
- systemctl restart systemd-journald
- docker run -it --rm kubernetes/pause

Signed-off-by: Jhon Honce jhonce@redhat.com

### CODE REVIEW

**Comment by runcom** [comment]
Ping @LK4D4 

LGTM

**Comment by thaJeztah** [comment]
We're currently not on Go 1.6 yet (we were, but reverted to 1.5.4, due to incompatibilities with older clients, see  https://github.com/docker/docker/issues/20865), so does this work?

**Comment by runcom** [comment]
This is basically a no-op on Go < 1.6

**Comment by cpuguy83** [comment]
This is pretty gross.

**Comment by runcom** [comment]
@cpuguy83 any other idea how the same could be done (ignoring SIGPIPE and docker daemon kill)

**Comment by runcom** [comment]
This has been worked based on the linked issue and comment https://github.com/docker/docker/issues/19728#issuecomment-192044258

googling "ignoring signals in golang" doesn't give us any more option in doing so - if anyone has an idea..

**Comment by LK4D4** [comment]
It is gross indeed, but silent daemon crash is gross too.
Design lgtm

File: cmd/dockerd/docker.go [hunk 1]
```diff
@@ -23,6 +25,17 @@ func main() {
 		return
 	}
 
+	// Fixes https://github.com/docker/docker/issues/19728
+	// Ignore SIGPIPE events, that are generated by systemd when journald is restarted while
+	// the docker daemon is not restarted and also running under systemd.
+	go func() {
+		c := make(chan os.Signal, 2048)
+		signal.Notify(c, os.Signal(syscall.SIGPIPE))
```
Comment (LK4D4) [inline cmd/dockerd/docker.go]: just `for range c`

File: cmd/dockerd/docker.go [hunk 2]
```diff
@@ -23,6 +25,17 @@ func main() {
 		return
 	}
 
+	// Fixes https://github.com/docker/docker/issues/19728
+	// Ignore SIGPIPE events, that are generated by systemd when journald is restarted while
+	// the docker daemon is not restarted and also running under systemd.
+	go func() {
+		c := make(chan os.Signal, 2048)
```
Comment (LK4D4) [inline cmd/dockerd/docker.go]: No need in buffered channel imo.

**Comment by LK4D4** [comment]
Also I think it's better to write short commit message title and move whole text to commit message body.

**Comment by runcom** [comment]
@jwhonce 
```
19:04:58 ---> Making bundle: validate-gofmt (in bundles/1.12.0-dev/validate-gofmt)
19:04:58 These files are not properly gofmt'd:
19:04:58  - cmd/dockerd/docker.go
19:04:58 
19:04:58 Please reformat the above files using "gofmt -s -w" and commit the result.
```

**Comment by justincormack** [comment]
SIGPIPE was a terrible idea. NetBSD has the sane `O_NOSIGPIPE` flag to `open`/`fcntl` which I think the other BSDs are adopting, but Linux does not annoyingly. This seems like the best workaround.

File: cmd/dockerd/docker.go [hunk 3]
```diff
@@ -23,6 +25,16 @@ func main() {
 		return
 	}
 
+	// Fixes https://github.com/docker/docker/issues/19728
+	// Ignore SIGPIPE events, that are generated by systemd when journald is restarted while
+	// the docker daemon is not restarted and also running under systemd.
+	go func() {
+		c := make(chan os.Signal)
+		signal.Notify(c, os.Signal(syscall.SIGPIPE))
+		for range c {
```
Comment (LK4D4) [inline cmd/dockerd/docker.go]: I still think that we should log them, on WARN level maybe.

Comment (jwhonce) [inline cmd/dockerd/docker.go]: @LK4D4 I didn't log any message to handle the use case where the logging is going to stdout/stderr.  Logging here in that case would create an infinite loop.

Comment (jwhonce) [inline cmd/dockerd/docker.go]: Just to be clear if you log a message in this handler:

Docker catches the SIGPIPE, would log to stdout/stderr, systemd would generate a SIGPIPE event, and the whole cycle repeats.

**Comment by crosbymichael** [comment]
You move this here:
https://github.com/docker/docker/blob/master/pkg/signal/trap.go#L25

we don't want to have multiple signal handlers scattered throughout the code base. Its better to keep them all in one place.

**Comment by jwhonce** [comment]
@crosbymichael I can move this code to signal.Trap() that would also require the call to signal.Trap() to be called earlier in daemon.start().  Otherwise, one of the logrus.Warn/Info calls currently in start() would cause a SIGPIPE event to be generated.

**Comment by icecrime** [comment]
@crosbymichael @LK4D4 PTAL!

File: pkg/signal/trap.go [hunk 4]
```diff
@@ -18,38 +18,43 @@ import (
...
-				logrus.Infof("Processing signal '%v'", sig)
-				switch sig {
-				case os.Interrupt, syscall.SIGTERM:
-					if atomic.LoadUint32(&interruptCount) < 3 {
-						// Initiate the cleanup only once
-						if atomic.AddUint32(&interruptCount, 1) == 1 {
-							// Call the provided cleanup handler
-							cleanup()
-							os.Exit(0)
+				if sig != syscall.SIGPIPE {
```
Comment (LK4D4) [inline pkg/signal/trap.go]: check it outside goroutine:
```
if sig == syscall.SIGPIPE {
    continue
}
```

**Comment by jwhonce** [comment]
@LK4D4 I implemented your suggestion

File: pkg/signal/trap.go [hunk 5]
```diff
@@ -18,15 +18,22 @@ import (
...
 	c := make(chan os.Signal, 1)
-	// we will handle INT, TERM, QUIT here
-	signals := []os.Signal{os.Interrupt, syscall.SIGTERM, syscall.SIGQUIT}
+	// we will handle INT, TERM, QUIT, SIGPIPE here
+	signals := []os.Signal{os.Interrupt, syscall.SIGTERM, syscall.SIGQUIT, syscall.SIGPIPE}
 	gosignal.Notify(c, signals...)
 	go func() {
 		interruptCount := uint32(0)
 		for sig := range c {
+			if sig != syscall.SIGPIPE {
```
Comment (LK4D4) [inline pkg/signal/trap.go]: I think it should be `sig == syscall.SIGPIPE`.

**Comment by thaJeztah** [comment]
ping @crosbymichael @LK4D4 PTAL, looks like your comments were addressed; is this ready to merge?

**Comment by LK4D4** [comment]
LGTM

**Comment by runcom** [comment]
LGTM

**Comment by thaJeztah** [comment]
Looks like we have two LGTM's and all green, so merging

**Pull request merged**